web-client/src/schema/crypto.ts

```diff
@@ -29,15 +38,97 @@ export class TweetNaClCrypto {
   static new = (): TweetNaClCrypto => new TweetNaClCrypto(nacl.box.keyPair());
 
   decrypt_message = (
-    ciphertext: Bytes,
+    ciphertext: Bytes32,
     their_pk: PublicKey,
     nonce: Nonce
   ): Uint8Array | null =>
     nacl.box.open(ciphertext, nonce, their_pk, this.secret_key);
 
-  encrypt_message = (message: Bytes, their_pk: PublicKey): EncryptedMessage => {
+  encrypt_message = (
+    message: Bytes32,
+    their_pk: PublicKey
+  ): EncryptedMessage => {
     const nonce = nacl.randomBytes(nacl.box.nonceLength);
     const ciphertext = nacl.box(message, nonce, their_pk, this.secret_key);
     return { ciphertext, nonce };
   };
 }
+
+/**
+ * A shared secret acquired via a DH key agreement
+ */
+export type SharedSecret = Bytes32;
+/**
+ * A public salt value
+ */
+export type Salt = Bytes;
+/**
+ * A seed used to instantiate a random number generator such as when generating
+ * a random key pair.
+ */
+export type Seed = Bytes32;
+
+type KeyAgreementFinished = boolean;
+export class DiffieHellman {
+  protected finished: KeyAgreementFinished = false;
+  protected hkdfSalt: Salt = new Uint8Array(0);
+  protected ourPk?: PublicKey;
+  protected x25519: X25519KeyAgreement = new X25519KeyAgreement();
+
+  protected constructor(hkdfSalt: Salt, seed?: Bytes32);
+  protected constructor(seed?: Bytes32) {
+    this.x25519 = new X25519KeyAgreement(seed);
+  }
+
+  /**
+   * Instantiate a new key agreement using a keypair generated from a seed.
+   * Generally, this method should be avoided and `new` should be used in its
+   * place.
+   */
+  static from_seed = (
+    seed: Bytes32,
+    salt: Salt = new Uint8Array(0)
+  ): DiffieHellman => new DiffieHellman(seed, salt);
+
+  /**
+   * Instantiate a new key agreement using a randomly generated keypair.
+   */
+  static new = (salt: Salt = new Uint8Array(0)): DiffieHellman =>
+    new DiffieHellman(salt);
+
+  /**
+   * Derive a new key pair and return the public key.  Otherwise, return a
+   * previously cached public key.
+   */
+  x25519_public_key = (): PublicKey => {
+    if (this.ourPk === undefined) {
+      this.ourPk = this.x25519.offer();
+    }
+    return this.ourPk;
+  };
+
+  /**
+   * Complete the ECDH operation in order to obtain a shared secret.
+   */
+  diffie_hellman = (
+    secret_length: number,
+    their_pk: PublicKey,
+    salt?: Bytes32
+  ): SharedSecret => {
+    if (salt === undefined) {
+      salt = new Uint8Array(0);
+    }
+    this.finish(their_pk);
+    const raw_shared_key: SharedSecret = this.x25519.getSharedKey();
+
+    // RFC 7748 recommends applying a KDF to the raw shared key
+    const hkdf = new HKDF(SHA256, raw_shared_key, salt);
```

Comment:
No.  The Salt is defined by us.  For now, it is sufficient to use the wallet id as the salt.  It might make sense to revisit that decision at a later point but this is robust enough for now.

Reply:
HKDF is a hash-based key derivation function.  Specifically, we are using `HKDF-SHA-256`.  It is an IETF standard (i.e. an RFC) for deriving keys using `HMAC-SHA-256`.